bindings/rust/evmc-declare/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+#![recursion_limit = "128"]
+
+extern crate proc_macro;
+
+use heck::ShoutySnakeCase;
+use heck::SnakeCase;
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::parse_macro_input;
+use syn::spanned::Spanned;
+use syn::Ident;
+use syn::IntSuffix;
+use syn::ItemStruct;
+use syn::Lit;
+use syn::LitInt;
+use syn::LitStr;
+use syn::Meta;
+use syn::MetaList;
+use syn::NestedMeta;
+
+#[proc_macro_attribute]
+pub fn evmc_declare_vm(args: TokenStream, item: TokenStream) -> TokenStream {
+    // First, try to parse the input token stream into an AST node representing a struct
+    // declaration.
+    let input: ItemStruct = parse_macro_input!(item as ItemStruct);
+
+    // Extract the identifier of the struct from the AST node.
+    let vm_type_name: String = input.ident.to_string();
+
+    // Get the name in shouty snake case for the statically defined VM data.
+    let vm_name_allcaps: String = vm_type_name.to_shouty_snake_case();
+
+    // Get the name in snake case and strip the underscores for the symbol name.
+    let vm_name_lowercase: String = vm_type_name
+        .to_snake_case()
+        .chars()
+        .filter(|c| *c != '_')
+        .collect();
+
+    // Parse the attribute meta-items for the name and version. Verify param count.
+    let meta = parse_macro_input!(args as MetaList);
+    assert!(
+        meta.nested.len() == 1,
+        "Incorrect number of meta-items passed to evmc_declare_vm."
+    );
+
+    // Extract the name and capabilities from the attribute's key-value item.
+    let (vm_name_stylized, vm_capabilities) = match &meta.nested[0] {
+        NestedMeta::Meta(m) => {
+            // Try to form a string from the identifier if a meta-item was supplied.
+            if let Meta::NameValue(pair) = m {
+                let name = pair.ident.to_string();
+                let capabilities = match pair.lit {
+                    Lit::Str(ref s) => match s.value().as_str() {
+                        "evm1" => 0x1u32,
+                        "ewasm" => 0x1u32 << 1u32,
+                        _ => panic!("Invalid capabilities specifier. Use 'evm1' or 'ewasm'."),                       
+                    },
+                    _ => panic!("Argument 1 is not a valid string literal."),
+                };
+                (name, capabilities)
+            } else {
+                panic!("Argument passed to evmc_declare_vm is not a name-value item.")
+            }
+        }
+        _ => panic!("Argument passed to evmc_declare_vm is of incorrect type. Expected a name-value item.")
+    };
+    
+    // Get the VM version from the crate version.
+    let vm_version_string = env!("CARGO_PKG_VERSION").to_string();
```

Comment:
While I like this we need to document it somehow.

Reply:
Yeah, not sure where to.